### ISSUE TITLE
(Android) Content bleeds into Indicator position due to height miscalculation

### ISSUE THREAD

**Issue by PauloPolaco** [alexrainman/CarouselView#185]
### Description of Defect
The _CarouselViewControl_ has an issue of not rendering the control's view height on _Android_ during first load. The control's content bleeds into and past the _Indicator_ which is situated at the bottom of the control. The bug only occurs when the control is on the primarily loaded page.  The bug does not occur during second loading of the page or when the control is situated on a secondary page.

### Bug Repro
Uncomment the _<xamFormsCarousel:CarouselViewControl>_, remove the _<c:Carousel>_ and run on the _Android_ version of the app.  Swipe left and right to see how the templates bleed over the bottom _Indicator_.  

[Clean repro project using MasterDetailPage as base](https://github.com/PauloPolaco/XamFormsCarouselFix)

### Fix
Removes the _CarouselViewControl's_ issue of not rendering the control's view height on _Android_ during first load.

### Additional Info
The problem might lye in the _Android_ renderer's _SetNativeView()_ function. It occurs regardless whether the _DataTemplate_ uses a _StackPanel_ or _Grid_ control as the wrapper.  Not sure about Windows as we use _UWP/WinRT_ solutions with MSFT based _XAML_ for that platform. This may possibly be related to the following:  

* [Android does not re-measure its layout on Height Changed](https://github.com/alexrainman/CarouselView/issues/184)  

* [Carousel miscalculating heights in some cases](https://github.com/alexrainman/CarouselView/issues/123)

**Comment by alexrainman** [comment]
It's known the Carousel has issues with MasterDetails page and StackLayout. Cna you describe what exactly is the issue you are facing?

**Comment by PauloPolaco** [comment]
As mentioned above, please take a look at my repro project.  Uncomment the _CarouselViewControl_ and comment out my derived _Carousel_ control from the MainView within the repro project.

[https://github.com/PauloPolaco/XamFormsCarouselFix](https://github.com/PauloPolaco/XamFormsCarouselFix)

Again, the issue is that when the _Element_SizeChanged()_ event gets called within your _Android_ renderer, it sets the _WidthElement_. However, the _HeightElement_ property does not get set before the event triggers the _SetNativeView()_ function for the first time.  Again, this issue occurs on first load only.  When the _CarouselViewControl_ is on the primary _Page_.

Please take a look at my derived _Android_ renderer.  I don't dispute that the actual issue lies within the _XamForms_ layout engine.  With that said, what I'm looking for is a native solution within the _CarouselViewControl_.  Rather than having to do a silly workaround like in my repro project.

![xamformscarouselfix](https://user-images.githubusercontent.com/9606406/28688824-435c4a74-72e1-11e7-93a5-32d64ad8ed17.png)

**Comment by alexrainman** [comment]
What im trying to tell you is that there's not actual fix for this. When you use the Carousel control with a master detail and stacklayout, you will have this problem. It happens with some other layout configurations though. If you read other users issues, they used workarounds.

**Comment by alexrainman** [comment]
Your native workaround works?

**Comment by PauloPolaco** [comment]
The native workaround to my issue would be to remove the following line:

```
ElementWidth = rect.Width;
```
From the _Element_SizeChanged()_ function within your _Android_ `CarouselViewRenderer`.  So it would read as such:

```
void Element_SizeChanged(object sender, EventArgs e)
{
	if (Element != null)
        {
       		// To avoid page recreation caused by entry focus #136 (fix)
                var rect = this.Element.Bounds;
                if (ElementWidth.Equals(rect.Width))
                    return;

                //ElementWidth = rect.Width;
                //ElementHeight = rect.Height;

                SetNativeView();
                Element.PositionSelected?.Invoke(Element, Element.Position);
	}
}
```

With that said.  I'm not sure what the actual ramifications of removing that would be.  Generally speaking and in terms of other users' existing projects.  That's why I never proposed it.  

If you look closely at the derived renderer within my repro project you can see how I masked your _Element_SizeChanged()_ event with my own and then called the private _SetNativeView()_ function from your base class via reflection.  This bypasses the _ElementWidth_ property from being set within your version of the event:

```
void Element_SizeChanged(object sender, EventArgs e)
{
	if (Element != null)
	{
		typeof(CarouselViewRenderer).GetMethod("SetNativeView",
			System.Reflection.BindingFlags.NonPublic |
			System.Reflection.BindingFlags.Instance).Invoke(
			(CarouselViewRenderer)this, null);

		Element.PositionSelected?.Invoke(Element, Element.Position);
	}
}
```

**Comment by alexrainman** [comment]
I just released a new version but a fix for this still not there as i have to handle a complicated scenario related to soft-keyboard.

**Comment by PauloPolaco** [comment]
Thank you for taking a look at this.  I had a hunch that my proposed fix would have some adverse effects.

Albeit somewhat cumbersome, my client side renderer fix within [Clean repro project using MasterDetailPage as base](https://github.com/PauloPolaco/XamFormsCarouselFix) still works a charm.

**Comment by EmmanVazz** [comment]
I have used @PauloPolaco solution and it also works for me.

**Comment by hardcodet** [comment]
@PauloPolaco I'm seeing this occasionally on iOS as well with a page that I display dynamically (added after the carousel was created). Do you think the issue you fixed is a general one, or Android specific?

**Comment by PauloPolaco** [comment]
@hardcodet The issue I had was with Android.  In my case, I'm loading everything within the CTOR, proir to initial page display.

Although I do remember seeing the same thing pop up for both Droid & Touch when I loaded the carousel templates post CTOR.  Eg. during the OnAppearing() overload or my own OnLoad function (yes, I built my own).  As previously beat to death by @alexrainman the real issue here is with the XamForms layout system (which is horrible) calculating the control and page heights differently during the CTOR (when the XAML is created) and only giving you the correct heights when everything is loaded.
